src/dataflow/src/source/mod.rs

```diff
@@ -736,29 +736,32 @@ pub struct PartitionMetrics {
     messages_ingested: DeleteOnDropCounter<'static, AtomicI64>,
     logger: Option<Logger>,
     source_name: String,
-    source_id: String,
+    source_id: SourceInstanceId,
     partition_id: String,
     last_offset: i64,
+    last_timestamp: i64,
 }
 
 impl PartitionMetrics {
     /// Record the latest offset ingested high-water mark
-    pub fn record_offset(&mut self, offset: i64) {
+    pub fn record_offset(&mut self, offset: i64, timestamp: i64) {
         if let Some(logger) = self.logger.as_mut() {
             logger.log(MaterializedEvent::SourceInfo {
                 source_name: self.source_name.clone(),
-                source_id: self.source_id.clone(),
+                source_id: self.source_id,
                 partition_id: self.partition_id.clone(),
                 offset: offset - self.last_offset,
+                timestamp: timestamp - self.last_timestamp,
```

Comment:
should we check that this number is positive? In certain settings we might not get timestamps in increasing order